usr/src/man/man1/gettext.1
'\" te
.\" Copyright (c) 2001, Sun Microsystems, Inc.  All Rights Reserved
.\" Copyright (c) 2012-2020, J. Schilling
.\" Copyright (c) 2013, Andreas Roehler
.\" CDDL HEADER START
.\"
.\" The contents of this file are subject to the terms of the
.\" Common Development and Distribution License ("CDDL"), version 1.0.
.\" You may only use this file in accordance with the terms of version
.\" 1.0 of the CDDL.
.\"
.\" A full copy of the text of the CDDL should have accompanied this
.\" source.  A copy of the CDDL is also available via the Internet at
.\" http://www.opensource.org/licenses/cddl1.txt
.\"
.\" When distributing Covered Code, include this CDDL HEADER in each
.\" file and include the License file at usr/src/OPENSOLARIS.LICENSE.
.\" If applicable, add the following below this CDDL HEADER, with the
.\" fields enclosed by brackets "[]" replaced with your own identifying
.\" information: Portions Copyright [yyyy] [name of copyright owner]
.\"
.\" CDDL HEADER END
.TH gettext 1 "27 Feb 2020" "SunOS 5.11" "User Commands"
.SH NAME
gettext, ngettext \- retrieve text string from message database
.SH SYNOPSIS
.LP
.nf
.BR gettext " [" "-d\c
.I " textdomain\c"
 | \c
.BR -\(midomain =\c
.I textdomain\c
]
.RI "     [" textdomain ] " msgid
.fi

.LP
.nf
.B gettext -s \c
.RB [ -e "] [" -n ]
.RB "     [" "-d \c"
.I textdomain\c
 | \c
.BR -\(midomain =\c
.IR textdomain "] \c"
.IR msgid ...
.fi

.LP
.nf
.BR ngettext " [" "-d\c
.I " textdomain\c"
 | \c
.BR -\(midomain =\c
.I textdomain\c
]
.RI "     [" textdomain ] " msgid msgid-plural count
.fi

.SH DESCRIPTION
.sp
.LP
The
.B gettext
utility retrieves a translated text string corresponding
to string
.I msgid
from a message object generated with
.BR msgfmt (1).
The message object name is derived from the optional argument
.I textdomain
if present, otherwise from the
.B TEXTDOMAIN
environment. If no domain is specified, or if a corresponding string cannot
be found,
.B gettext
prints
.IR msgid .
.sp
.LP
The
.B ngettext
utility retrieves a translated text string corresponding
to string
.I msgid
or
.I msgid-plural
where
.I count
determines if and which plural form is to be used
from a message object generated with
.BR msgfmt (1).
The message object name is derived from the optional argument
.I textdomain
if present, otherwise from the
.B TEXTDOMAIN
environment.
If no domain is specified, or if a corresponding string cannot be found
.B ngettext
prints
.I msgid
if count == 1 or
.I msgid-plural
if count != 1.

.sp
.LP
Ordinarily,
.B gettext
and
.B ngettext
look for their message object in
.BI /usr/lib/locale/ lang\c
/LC_MESSAGES
where
.I lang
is the locale
name. If present, the
.B TEXTDOMAINDIR
environment variable replaces the
pathname component up to
.IR lang .
.sp
.LP
This command interprets the C escape sequences 
.BR \eb ,
.BR \ef ,
.BR \en ,
.BR \er ,
.BR \et ,
.BR \ev ,
.BR \e\e .
Use
.B \et
to print a
.BR tab ,
use
.B \e\e
to print a
.BR backslash .
To produce a message on a line of its own,
either enter
.B \en
at the end of
.IR msgid ,
or use this command in
conjunction with
.BR printf (1).
.sp
.LP
When used with the
.B \-s
option,
.B gettext
behaves like
.BR echo (1),
except that escape sequences are not expanded unless the
.B \-e
option is specified as well.
It however does not simply copy its arguments to standard output.
Instead, those messages found in the selected catalog are translated.
.SH OPTIONS
.sp
.LP
The following options are supported:
.sp
.ne 2
.TP 26n
.PD 0
.BI -d " textdomain
.TP
.BI -\(midomain\fR= textdomain
.PD
Retrieves translated messages from the domain
.IR textdomain ,
if
.I textdomain
is not specified as an operand.

.sp
.ne 2
.TP
.B -e
Enables expansion of the escape sequences
.BR \eb ,
.BR \ef ,
.BR \en ,
.BR \er ,
.BR \et ,
.BR \ev ,
.B \e\e
in the untranslated text if used with the
.BR -s
option.

.sp
.ne 2
.TP
.B -n
Suppresses trailing newline if used with the
.B -s
option.

.sp
.ne 2
.TP
.B -s
Behaves like
.BR echo (1)
(see DESCRIPTION above). If the
.B -s
option
is specified, no expansion of C escape sequences is performed and a newline
character is appended to the output, by default.

.SH OPERANDS
.sp
.LP
The following operands are supported:
.sp
.ne 2
.TP 14n
.I textdomain
A domain name used to retrieve the messages. This overrides the
specification by the
.B -d
or
.B -\(midomain
options, if present.

.sp
.ne 2
.TP
.I msgid
A key to retrieve the localized message.

.SH ENVIRONMENT VARIABLES
.sp
.ne 2
.TP 17n
.B LANG
Specifies locale name.

.sp
.ne 2
.TP 17n
.B LANGUAGE
is a
list of locale names separated by a colon (':') character.  If
.B LANGUAGE
is defined, each locale name is tried in the specified order and if a
GNU-compatible message catalogue is found, the message is returned.  If a
GNU-compatible message catalogue is found but failed to find a corresponding
.IR msgid ,
the
.I msgid
string is returned. If
.B LANGUAGE
is not
defined or if a Solaris message catalogue is found or no GNU-compatible
message catalogue is found in processing
.BR LANGUAGE ,
the pathname used to
locate the message catalogue is
.IR dirname / locale / category /\c
.IR domainname .mo,
where
.I dirname
is the directory specified by
.BR TEXTDOMAINDIR ,
.I locale
is a locale name, and
.I category
is
.BR LC_MESSAGES .

.sp
.ne 2
.TP
.B LC_MESSAGES
Specifies messaging locale, and if present overrides
.B LANG
for
messages.

.sp
.ne 2
.TP
.B NLSPATH
The
.B NLSPATH
environment variable (see
.BR environ (5))
is searched
first for the location of the
.B LC_MESSAGES
catalogue. The setting of the
.B LC_MESSAGES
category of the current locale determines the locale used
for string retrieval.
If
.B NLSPATH
is not defined and the current locale is "C",
.B gettext
simply returns the message string that was passed. 
In a locale other than "C", if
.B NLSPATH
is not defined or
if a message catalogue is not found in any of the components specified by
.BR NLSPATH ,
the routines search for the message catalogue using the scheme
described in the
.B LANGUAGE
description.

.sp
.ne 2
.TP
.B TEXTDOMAIN
Specifies the text domain name, which is identical to the message object
filename without
.B \&.mo
suffix.

.sp
.ne 2
.TP
.B TEXTDOMAINDIR
Specifies the pathname to the message database. If present, replaces
.BR /usr/lib/locale .

.SH ATTRIBUTES
.sp
.LP
See
.BR attributes (5)
for descriptions of the following attributes:
.sp

.SH "EXIT STATUS
.LP
The following error values are returned:
.sp
.TP
.B 0
Successful completion.
.TP
.B !=0
An error occurred.

.sp
.TS
tab() box;
cw(2.75i) |cw(2.75i)
lw(2.75i) |lw(2.75i)
.
ATTRIBUTE TYPEATTRIBUTE VALUE
_
AvailabilitySUNWcsu
.TE

.SH SEE ALSO
.sp
.LP
.BR echo (1),
.BR msgfmt (1),
.BR printf (1),
.BR gettext (3C),
.BR setlocale (3C),
.BR attributes (5)
.SH NOTES
.sp
.LP
This is the shell equivalent of the library routine
.BR gettext (3C).


usr/src/cmd/gettext/gettext.c
/*
 * CDDL HEADER START
 *
 * The contents of this file are subject to the terms of the
 * Common Development and Distribution License, Version 1.0 only
 * (the "License").  You may not use this file except in compliance
 * with the License.
 *
 * A full copy of the text of the CDDL should have accompanied this
 * source.  A copy of the CDDL is also available via the Internet at
 * http://www.opensource.org/licenses/cddl1.txt
 *
 * When distributing Covered Code, include this CDDL HEADER in each
 * file and include the License file at usr/src/OPENSOLARIS.LICENSE.
 * If applicable, add the following below this CDDL HEADER, with the
 * fields enclosed by brackets "[]" replaced with your own identifying
 * information: Portions Copyright [yyyy] [name of copyright owner]
 *
 * CDDL HEADER END
 */
/*
 * Copyright 2020 J. Schilling
 * Copyright 2004 Sun Microsystems, Inc.  All rights reserved.
 * Use is subject to license terms.
 */

/*	Copyright (c) 1983, 1984, 1985, 1986, 1987, 1988, 1989 AT&T	*/
/*	  All Rights Reserved  	*/

/*
 * Portions of this source code were derived from Berkeley 4.3 BSD
 * under license from the Regents of the University of California.
 */

#pragma ident	"%Z%%M%	%I%	%E% SMI"

#include <stdlib.h>
#include <unistd.h>
#include <stdio.h>
#include <string.h>
#include <locale.h>
#include <libgen.h>	/* basename() */

/*
 * TEXTDOMAIN should be defined in Makefile
 * in case it isn't, define it here
 */
#if !defined(TEXT_DOMAIN)
#define	TEXT_DOMAIN "SYS_TEST"
#endif

static char *
expand_metas(char *in)	/* walk thru string interpreting \n etc. */
{
	register char *out, *cp;

	for (cp = out = in; *in != NULL; out++, in++) {
		if (*in == '\\') {
			switch (*++in) {
			case 'b' :
				*out = '\b';
				break;
			case 'f' :
				*out = '\f';
				break;
			case 'n' :
				*out = '\n';
				break;
			case 'r' :
				*out = '\r';
				break;
			case 't' :
				*out = '\t';
				break;
			case 'v' :
				*out = '\v';
				break;
			default:
				*out = *in;
				break;
			}
		} else
			*out = *in;
	}
	*out = NULL;
	return (cp);
}

#define	ERR_USAGE \
	"Usage: gettext [-d domainname | --domain=domainname ]  " \
		"[domain] \"msgid\"\n" \
	"       gettext -s [-d domainname | --domain=domainname] [-e] [-n] "\
		"\"msgid\" ...\n"

#define	ERR_NUSAGE \
	"Usage: ngettext [-d domainname | --domain=domainname ]  " \
		"[domain] \"msgid\" \"msgid-plural\" count\n"

static void
usage(int is_ngettext) {
	if (is_ngettext)
		(void) fprintf(stderr, gettext(ERR_NUSAGE));
	else
		(void) fprintf(stderr, gettext(ERR_USAGE));
	exit(-1);
}

int
main(int argc, char *argv[])	/* shell script equivalent of gettext(3) */
{
	char	*domainpath, *msgid;
	char	*domain = NULL;
	char	c, *arg;
	int	exp_flag = 0;
	int	no_newline = 0;
	int	echo_flag = 0;
	int	is_ngettext = 0;

	(void) setlocale(LC_ALL, "");
	(void) textdomain(TEXT_DOMAIN);

	arg = basename(argv[0]);
	if (*arg == 'n')
		is_ngettext = 1;
	argv++;
	while (--argc > 1) {
		arg = *argv;
		if (*arg == '-') {
			if (!*(arg + 1)) {
				/* not an option */
				break;
			}
loop:
			if ((c = *++arg) == '\0') {
				/* next argument */
				argv++;
				continue;
			} else if (c != '-') {
				switch (c) {
				case 'd':
					/* domainname */
					if (*(arg + 1)) {
						/*
						 * no spaces between -d and
						 * optarg
						 */
						domain = ++arg;
						argv++;
						continue;
					}
					if (--argc > 1) {
						domain = *++argv;
						argv++;
						continue;
					}
					/* not enough args */
					usage(is_ngettext);
					/* NOTREACHED */
					break;
				case 'e':
					/* enable escape sequence expansion */
					exp_flag = 1;
					goto loop;
					/* NOTREACHED */
				case 'n':
					/* suppress tailing newline */
					no_newline = 1;
					if (is_ngettext == 0)
						goto loop;
					/* FALLTHROUGH */
				case 's':
					echo_flag = 1;
					if (is_ngettext == 0)
						goto loop;
					/* FALLTHROUGH */
				default:
					/* illegal option */
					usage(is_ngettext);
					/* NOTREACHED */
					break;
				}
				/* NOTREACHED */
			}
			/* c == '-' */
			if (*(arg + 1) == '\0') {
				/* "--" found, option end */
				argv++;
				argc--;
				break;
			}

			/* long option */
			arg++;
			if (strncmp(arg, "domain=", 7) == 0) {
				/* domainname */
				arg += 7;
				if (*arg == '\0') {
					/* illegal option */
					usage(is_ngettext);
					/* NOTREACHED */
				}
				domain = arg;
				argv++;
				continue;
			}
			/* illegal option */
			usage(is_ngettext);
			/* NOTREACHED */
		}
		break;
	}
	if ((argc == 0) ||
	    (is_ngettext && argc < 3) ||
	    (!is_ngettext && argc > 2)) {
		usage(is_ngettext);
		/* NOTREACHED */
	}

	domainpath = getenv("TEXTDOMAINDIR");
	if (!echo_flag) {
		/* traditional mode */
		if ((!is_ngettext && argc == 2) ||
		   (is_ngettext && argc == 4)) {
			/*
			 * textdomain is specified by the argument.
			 */
			domain = *argv++;
		} else if (!domain) {
			/*
			 * textdomain is not specified by the argument.
			 * TEXTDOMAIN will be used.
			 */
			domain = getenv("TEXTDOMAIN");
			if (!domain) {
				/*
				 * no domain specified
				 * Just print the argument given.
				 */
				if (is_ngettext && atol(argv[2]) != 1) {
					(void) printf("%s",
							expand_metas(argv[1]));
					exit(1);
				}
				(void) printf("%s", expand_metas(*argv));
				exit(1);
			}
		}
		if (domainpath) {
			(void) bindtextdomain(domain, domainpath);
		}
		msgid = expand_metas(*argv);
		if (is_ngettext) {
			char		*msgid2 = expand_metas(argv[1]);
			unsigned long	n = atol(argv[2]);

			(void) fputs(dngettext(domain, msgid, msgid2, n),
					stdout);
		} else {
			(void) fputs(dgettext(domain, msgid), stdout);
		}
		exit(*domain == NULL);
	}
	/* echo mode */
	if (!domain) {
		domain = getenv("TEXTDOMAIN");
	}
	if (domainpath && domain) {
		(void) bindtextdomain(domain, domainpath);
	}
	while (argc-- > 0) {
		if (exp_flag)
			msgid = expand_metas(*argv++);
		else
			msgid = *argv++;
		(void) fputs(domain ? dgettext(domain, msgid) : msgid,
			stdout);

		if (argc > 0)
			(void) fputc(' ', stdout);
	}
	if (!no_newline)
		(void) fputc('\n', stdout);

	return ((domain == NULL) || (*domain == NULL));
}
